### PR TITLE
1785 Add export button and notification dialogs

### LINKED ISSUE #1785: Implement Latest Export Button UI
### Overview

We need to implement UI elements that allow users to export selected data on to their devices since we do not officially have a way of utilizing our data export feature.

### Note

Handoff documents will contain a version of the Search and Filters modal that DO NOT MATCH what we have implemented. We are to implement only the UI elements relevant to this feature
- [x] I understand

### Action Items

- [x] Add export button at bottom of Search and Filters Modal
- [x] Error modal is displayed on Export Button click when criteria is not met
  - [x] More than 1 request type is selected 
  - [x] No request types selected
  - [x] No NC is selected
- [x] NC selected + 1 request type selected begins data export on Export Button click
  - [x] data downloading modal appears
  - [x] data download success modal appears if data download succeeds
  - [x] data download failed modal appears if download failed

### Resources/Instructions

- [Link To Figma](https://www.figma.com/design/eF63O8yIGz2soQ4JJFfyoK/311_Data_UXUI?node-id=4081-9245)
- Page Name: ⭐ ⭐ ⭐ Handoff 
- Section Name: `Export Link Button And Hand Off Process`
- Related previous ticket: https://github.com/hackforla/311-data/issues/1695
  - [x] review: Before Screenshot
  - [x] review: After Screenshot (Finalized)
  - [x] review: Video Upload of "Export Button" demo 

<details><summary>Screenshot of Figma Overview</summary>
<p>

image

</p>
</details> 
<details><summary>Screenshot after implemented changes</summary>
<p>

image

image

image

image

image

image

image
</p>
</details>

### PR DESCRIPTION
**Pull request by kdow**
Fixes #1785 

  - [x] Up to date with `main` branch
  - [x] Branch name follows [guidelines](https://github.com/hackforla/311-data/blob/master/GETTING_STARTED.md#feature-branching)
  - [x] All PR Status checks are successful
  - [x] Peer reviewed and approved

Any questions? See the [getting started guide](https://github.com/hackforla/311-data/blob/master/GETTING_STARTED.md)

### CODE REVIEW

**Comment by traycn** [comment]
Hey @kdow I pulled the branch and it's running great on my end, just came across a minor issue with MUI.

For functionality, I tested the export feature with multiple requests with each request containing a archive folder of 2 files: NeighborhoodData.csv and ServiceRequestCount.csv. It's working great!

On the UI side of things, I get a console error for having the onHandler function exec in the Button element:

1785_console_error

There is an IconButton element in MUI that the error suggests we use. Links to the doc and examples below. What is your bandwidth for changing the Button element to IconButton as a potential fix for this console error?

IconButton Doc & Examples:
https://mui.com/material-ui/api/icon-button/
https://mui.com/material-ui/api/icon-button/

PS appreciate all the time put into this!

**Comment by kdow** [comment]
> On the UI side of things, I get a console error for having the onHandler function exec in the Button element:
> 1785_console_error
> 
> There is an IconButton element in MUI that the error suggests we use. Links to the doc and examples below. What is your bandwidth for changing the Button element to IconButton as a potential fix for this console error?
> 
> IconButton Doc & Examples: https://mui.com/material-ui/api/icon-button/

Thanks for catching that. It looks like it was referencing an IconButton component I created, and I just needed to move the onClick handler into it from the child Icon component. I've tested and it looks like the error went away. I've just updated the PR with the fix.

**Comment by traycn** [comment]
Thanks! Pulled the changes and came across another small issue. The functionality works great, just another MUI issue from a type check.

1785_console_error2

Expecting that it's being thrown at the following:
components/common/ChipList/StyledChip.jsx ---  line 84
`label: PropTypes.string.isRequired,`

**Comment by kdow** [comment]
Fixed!

**Comment by traycn** [comment]
LGTM thank you!

**Review by traycn — APPROVED** [review]

File: components/main/Desktop/Export/ExportButton.jsx [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+          }
+        }
+      };
+
+      if (selectionValidated) {
+        exportData();
+      }
+    };
+
+    getDataToExport();
```
Comment (ryanfchase) [inline components/main/Desktop/Export/ExportButton.jsx:205]: Alright I think I understand how this is working. No action needed, unless I've got any of the details wrong in the summary below:

In the previous version of this file, Export button's on-click behavior would generate and download the zip file (no validations were necessary at the time). In this version of the file, we now have the following state that gets modified when determining if a selection was validated:
- requestType: needed for passing along to export dialog
- selectionValidated: determined based on validation criteria
- setExportConfirmed: determined if the user confirmed their export selection
- dialogType: determined based on validation criteria

IFF the validation criteria is met at the time of clicking the Export button, then we allow the UseEffect to utilize the db connection, load the required data, and trigger the dialog prompting the user to confirm their export selection.

**Review by ryanfchase — CHANGES_REQUESTED** [review]
Minor changes requested, but overall this is great work. Ty @kdow 👍

File: components/main/Desktop/Export/ExportButton.jsx [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+import useStyles from './useStyles';
+import ExportDialog from './ExportDialog';
+
+function ExportButton({ filters }) {
+  const { conn } = useContext(DbContext);
+  const classes = useStyles();
+  const [showDialog, setShowDialog] = useState(false);
+  const [errorType, setErrorType] = useState('');
+  const [requestType, setRequestType] = useState('');
+  const [selectionValidated, setSelctionValidated] = useState(false);
```
[the remark this answers is not in the document]
Comment (kdow) [reply components/main/Desktop/Export/ExportButton.jsx]: fixed

File: components/main/Desktop/Export/ExportButton.jsx [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+      if (exportConfirmed) {
+        setDialogType('downloading');
+        try {
+          saveAs(content, '311Data.zip');
+          setDialogType('success');
+        } catch {
+          setDialogType('failed');
+        } finally {
+          setShowDialog(true);
```
[the remark this answers is not in the document]
Comment (kdow) [reply components/main/Desktop/Export/ExportButton.jsx]: Yeah, I think so. I've removed that line and tested it's working.

File: components/main/Desktop/Export/ExportButton.jsx (hunk 1 above)
Comment (kdow) [reply components/main/Desktop/Export/ExportButton.jsx:204]: Yes, this all is correct.

**Review by ryanfchase — APPROVED** [review]
New changes are looking good, approved. Merge when ready 👍

**Pull request merged**